README.md
## Raspberry Pi 2 Piglow (using the I2C skill)

NOTE: This repository is a work-in-progress. The I2C skill is neither merged
into snappy nor supported by Canonical. The repository serves as a way to share
ongoing work into this area.

# Installation

To install this snap on a 16.04 snappy device, run:

```
sudo snappy install pi2-piglow.zygoon
```

# Usage

Stay tuned, this currently requires code that is not merged into snappy. I will update usage instructions when it is possible to use on a vanilla installation of snappy.

# Demo

[![Demonstration video](https://i.vimeocdn.com/video/556514659.jpg?mw=1200&mh=675&q=70)](https://vimeo.com/155690419)


NEWS
Version 2.1:

- the .background application now reacts to SIGTERM sent by systemd and shuts
  down any glowing LEDs.

Version 2:

- The .demo application was replaced with the .foreground application.
- A new .background application was added to run the "breathe" animation
  in a never-ending loop.

Version 1:

- Initial release


piglow.c
#include "piglow.h"

#include "sn3218.h"

#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

static const uint8_t piglow_to_sn3218[18] = {
    6, 7, 8, 5, 4, 9, // first arm
    17, 16, 15, 13, 11, 10, // second arm
    0, 1, 2, 3, 14, 12 // third arm
};

int piglow_open(struct piglow *glow, const char *device_path) {
    glow->fd = open(device_path, O_RDWR);
    if (glow->fd < 0) {
        return errno;
    }
    return 0;
}

int piglow_close(struct piglow *glow) {
    if (close(glow->fd) < 0) {
        return errno;
    }
    glow->fd = -1;
    return 0;
}

int piglow_powerctl(struct piglow *glow, bool active) {
    return sn3218_powerctl(glow->fd, active);
}

int piglow_pwmctl(struct piglow *glow, const uint8_t pwm[18]) {
    uint8_t remapped_pwm[18];
    for (int i=0; i<18; ++i)
        remapped_pwm[piglow_to_sn3218[i]] = pwm[i];
    return sn3218_pwmctl(glow->fd, remapped_pwm);
}

int piglow_ledctl(struct piglow *glow, const bool active[18]) {
    bool remapped_active[18];
    for (int i=0; i<18; ++i)
        remapped_active[piglow_to_sn3218[i]] = active[i];
    return sn3218_ledctl(glow->fd, remapped_active);
}

int piglow_update(struct piglow *glow) {
    return sn3218_update(glow->fd);
}

int piglow_reset(struct piglow *glow) {
    return sn3218_reset(glow->fd);
}


sn3218.c
#include "sn3218.h"

#include <errno.h>
#include <sys/ioctl.h>
#include <linux/i2c.h>
#include <linux/i2c-dev.h>

int sn3218_powerctl(int fd, bool active) {

    struct i2c_rdwr_ioctl_data msgs = {
        .msgs = (struct i2c_msg[]){
            {
            .addr = SN3218_ADDR,
            .len = 2,
            .buf = (uint8_t[2]){
                SN3218_REG_SHUTDOWN,
                active
                    ? SN3218_MODE_NORMAL_OPERATION
                    : SN3218_MODE_SOFTWARE_SHUTDOWN},
            },
        },
        .nmsgs = 1,
    };
    if (ioctl(fd, I2C_RDWR, &msgs) < 0) {
        return errno;
    }
    return 0;
}

static const uint8_t gamma32[32] = {
    0, 1, 2, 4, 6, 10, 13, 18,
    22, 28, 33, 39, 46, 53, 61, 69,
    78, 86, 96, 106, 116, 126, 138, 149,
    161, 173, 186, 199, 212, 226, 240, 255,
};

static int min(int a, int b) {
    return a < b ? a : b;
}

static float interpolate(float v0, float v1, float t) {
      return (1 - t) * v0 + t * v1;
}

int sn3218_pwmctl(int fd, const uint8_t pwm[18]) {
    uint8_t adj_pwm[18];
    for (int i=0; i<18; ++i) {
        int level = pwm[i];
        int bin = level / 8;
        float t = level % 8 / 8.0;
        float v0 = gamma32[bin];
        float v1 = gamma32[min(31, bin + 1)];
        adj_pwm[i] = (uint8_t)interpolate(v0, v1, t);
    }
    struct i2c_rdwr_ioctl_data msgs = {
        .msgs = (struct i2c_msg[]){
            {
            .addr = SN3218_ADDR,
            .len = 1 + 18,
            .buf = (uint8_t[1 + 18]){
                SN3218_REG_PWM_BASE,
                adj_pwm[0], adj_pwm[1], adj_pwm[2],
                adj_pwm[3], adj_pwm[4], adj_pwm[5],
                adj_pwm[6], adj_pwm[7], adj_pwm[8],
                adj_pwm[9], adj_pwm[10], adj_pwm[11],
                adj_pwm[12], adj_pwm[13], adj_pwm[14],
                adj_pwm[15], adj_pwm[16], adj_pwm[17]},
            },
        },
        .nmsgs = 1,
    };
    if (ioctl(fd, I2C_RDWR, &msgs) < 0) {
        return errno;
    }
    return 0;
}

int sn3218_ledctl(int fd, const bool active[18]) {
    struct i2c_rdwr_ioctl_data msgs = {
        .msgs = (struct i2c_msg[]){
            {
            .addr = SN3218_ADDR,
            .len = 1 + 3,
            .buf = (uint8_t[1 + 3]){
                SN3218_REG_LED_CTL1,
                ( (active[0] ? (1 << 0) : 0) |
                  (active[1] ? (1 << 1) : 0) |
                  (active[2] ? (1 << 2) : 0) |
                  (active[3] ? (1 << 3) : 0) |
                  (active[4] ? (1 << 4) : 0) |
                  (active[5] ? (1 << 5) : 0)),
                ( (active[6] ? (1 << 0) : 0) |
                  (active[7] ? (1 << 1) : 0) |
                  (active[8] ? (1 << 2) : 0) |
                  (active[9] ? (1 << 3) : 0) |
                  (active[10] ? (1 << 4) : 0) |
                  (active[11] ? (1 << 5) : 0)),
                ( (active[12] ? (1 << 0) : 0) |
                  (active[13] ? (1 << 1) : 0) |
                  (active[14] ? (1 << 2) : 0) |
                  (active[15] ? (1 << 3) : 0) |
                  (active[16] ? (1 << 4) : 0) |
                  (active[17] ? (1 << 5) : 0))
                  },
            },
        },
        .nmsgs = 1,
    };
    if (ioctl(fd, I2C_RDWR, &msgs) < 0) {
        return errno;
    }
    return 0;
}

int sn3218_update(int fd) {
    struct i2c_rdwr_ioctl_data msgs = {
        .msgs = (struct i2c_msg[]){
            {
            .addr = SN3218_ADDR,
            .len = 2,
            // NOTE: 0xff is irrelevant
            .buf = (uint8_t[2]){SN3218_REG_UPDATE, 0xFF},
            },
        },
        .nmsgs = 1,
    };
    if (ioctl(fd, I2C_RDWR, &msgs) < 0) {
        return errno;
    }
    return 0;
}

int sn3218_reset(int fd) {
    struct i2c_rdwr_ioctl_data msgs = {
        .msgs = (struct i2c_msg[]){
            {
            .addr = SN3218_ADDR,
            .len = 2,
            // NOTE: 0xff is irrelevant
            .buf = (uint8_t[2]){SN3218_REG_RESET, 0xFF},
            },
        },
        .nmsgs = 1,
    };
    if (ioctl(fd, I2C_RDWR, &msgs) < 0) {
        return errno;
    }
    return 0;
}


sn3218.h
#pragma once
#include <stdint.h>
#include <stdbool.h>

// I2C address of the SN3218 chip.
#define SN3218_ADDR 0x54

// SN3218 shutdown register.
#define SN3218_REG_SHUTDOWN 0x00
// SN3218 PWM register base address (valid for + 0..17)
#define SN3218_REG_PWM_BASE 0x01
// SN32128 LED control register 1 (channel 1-6 enable bit).
#define SN3218_REG_LED_CTL1 0x13
// SN32128 LED control register 2 (channel 7-12 enable bit).
#define SN3218_REG_LED_CTL2 0x14
// SN32128 LED control register 3 (channel 13-18 enable bit).
#define SN3218_REG_LED_CTL3 0x15
// SN3218 update register.
#define SN3218_REG_UPDATE 0x16
// SN3218 state reset register.
#define SN3218_REG_RESET 0x17

// Argument for SN3218_REG_SHUTDOWN, puts the chip to soft-shutdown mode.
#define SN3218_MODE_SOFTWARE_SHUTDOWN 0x00
// Argument for SN3218_REG_SHUTDOWN, turns the chip on.
#define SN3218_MODE_NORMAL_OPERATION 0x01

// Control the power state of SN3218 chip.
int sn3218_powerctl(int fd, bool active);
// Control the PWM duty cycle of each of the 18 LEDs.
int sn3218_pwmctl(int fd, const uint8_t pwm[18]);
// Control the enable flag of each of the 18 LEDs.
int sn3218_ledctl(int fd, const bool active[18]);
// Commit values of PWM duty cycle and LED control registers.
int sn3218_update(int fd);
// Reset all state to initial values.
int sn3218_reset(int fd);


main.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <argp.h>
#include <string.h>

#include "piglow.h"

volatile sig_atomic_t shutdown = 0;

static void
on_shutdown_signal(int signo) {
    shutdown = 1;
}

const char *argp_program_version = "piglow 2.1";

/* Program documentation. */
static char doc[] =
"piglow -- example program using the piglow accessory on a Raspberry Pi 2";

/* A description of the arguments we accept. */
static char args_doc[] = "";

/* The options we understand. */
static struct argp_option options[] = {
    {"breathe",   'b',  0,  0, "Produce breathing pattern"},
    {"cycle",     'c',  0,  0, "Cycle through all LEDs"},
    {"forever",   'e',  0,  0, "Run forever!"},
    { 0 }
};

/* Used by main to communicate with parse_opt. */
struct arguments
{
    const char *mode;
    bool forever;
};

/* Parse a single option. */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
    /* Get the input argument from argp_parse, which we
       know is a pointer to our arguments structure. */
    struct arguments *arguments = state->input;

    switch (key)
    {
        case 'b':
            arguments->mode = "breathe";
            break;
        case 'c':
            arguments->mode = "cycle";
            break;
        case 'e':
            arguments->forever = true;
            break;
        case ARGP_KEY_ARG:
            if (state->arg_num >= 0)
                /* Too many arguments. */
                argp_usage (state);
            break;
        case ARGP_KEY_END:
            if (state->arg_num < 0)
                /* Not enough arguments. */
                argp_usage (state);
            break;
        default:
            return ARGP_ERR_UNKNOWN;
    }
    return 0;
}

/* Our argp parser. */
static struct argp argp = { options, parse_opt, args_doc, doc };

int main(int argc, char **argv) {
    struct piglow glow;
    bool led[18];
    uint8_t pwm[18];
    int err;
    struct arguments arguments = {
        .mode = "cycle,breathe",
        .forever = false,
    };

    argp_parse (&argp, argc, argv, 0, 0, &arguments);
    // TODO: learn about this through the snappy i2c skill.
    if ((err = piglow_open(&glow, "/dev/i2c-1"))) {
        fprintf(stderr, "Cannot open piglow: %s\n", strerror(err));
        goto fail;
    }
    signal(SIGINT, on_shutdown_signal);
    signal(SIGTERM, on_shutdown_signal);
    signal(SIGQUIT, on_shutdown_signal);
    // Initialize piglow
    printf("Resetting piglow\n");
    if ((err = piglow_reset(&glow))) {
        fprintf(stderr, "Cannot reset piglow: %s\n", strerror(err));
        goto fail;
    }
    printf("Powering piglow on\n");
    if ((err = piglow_powerctl(&glow, true))) {
        fprintf(stderr, "Cannot enable piglow: %s\n", strerror(err));
        goto fail;
    }
    do {
        if (shutdown)
            goto out;
        if (strstr(arguments.mode, "cycle") != NULL) {
            // Cycle through each LEDs at medium brigtness.
            printf("Cycling through each LED\n");
            for (int i=0; i<18; ++i) {
                led[i] = false;
                pwm[i] = 127;
            }
            if ((err = piglow_pwmctl(&glow, pwm))) {
                fprintf(stderr, "Cannot set PWM levels: %s\n", strerror(err));
                goto fail;
            }
            for (int i=0; i<18; ++i) {
                led[i] = true;
                if ((err = piglow_ledctl(&glow, led))) {
                    fprintf(stderr, "Cannot set LED control flags: %s\n", strerror(err));
                    goto fail;
                }
                led[i] = false;
                if ((err = piglow_update(&glow))) {
                    fprintf(stderr, "Cannot update piglow: %s\n", strerror(err));
                    goto fail;
                }
                usleep(200 * 1000);
                if (shutdown)
                    goto out;
            }
            for (int i=0; i<18; ++i) {
                led[i] = false;
            }
            if ((err = piglow_ledctl(&glow, led))) {
                fprintf(stderr, "Cannot set LED control flags: %s\n", strerror(err));
                goto fail;
            }
            if ((err = piglow_pwmctl(&glow, pwm))) {
                fprintf(stderr, "Cannot set PWM levels: %s\n", strerror(err));
                goto fail;
            }
            if ((err = piglow_update(&glow))) {
                fprintf(stderr, "Cannot update piglow: %s\n", strerror(err));
                goto fail;
            }
        }
        if (strstr(arguments.mode, "breathe") != NULL) {
            // Fade all LEDs together
            printf("Breathing...\n");
            for (int loop=0; loop<6; ++loop) {
                for (int i=0; i<18; ++i)
                    led[i] = i % 6 == loop;
                if ((err = piglow_ledctl(&glow, led))) {
                    fprintf(stderr, "Cannot set LED control flags: %s\n", strerror(err));
                    goto fail;
                }
                for (int br=0; br<=255; ++br) {
                    for (int i=0; i<18; ++i)
                        pwm[i] = br;
                    // Cycle through each LEDs at maximum brightness.
                    if ((err = piglow_pwmctl(&glow, pwm))) {
                        fprintf(stderr, "Cannot set PWM levels: %s\n", strerror(err));
                        goto fail;
                    }
                    if ((err = piglow_update(&glow))) {
                        fprintf(stderr, "Cannot update piglow: %s\n", strerror(err));
                        goto fail;
                    }
                    usleep(2000);
                    if (shutdown)
                        goto out;
                }
                usleep(1000 * 1000);
                if (shutdown)
                    goto out;
                for (int br=255; br>=0; --br) {
                    for (int i=0; i<18; ++i)
                        pwm[i] = br;
                    // Cycle through each LEDs at maximum brightness.
                    if ((err = piglow_pwmctl(&glow, pwm))) {
                        fprintf(stderr, "Cannot set PWM levels: %s\n", strerror(err));
                        goto fail;
                    }
                    if ((err = piglow_update(&glow))) {
                        fprintf(stderr, "Cannot update piglow: %s\n", strerror(err));
                        goto fail;
                    }
                    usleep(2000);
                    if (shutdown)
                        goto out;
                }
                usleep(1000 * 1000);
                if (shutdown)
                    goto out;
            }
        }
    } while (arguments.forever);
out:
    // Disable piglow
    if (shutdown) {
        printf("\n");
    }
    printf("Powering piglow off\n");
    if ((err = piglow_powerctl(&glow, false))) {
        fprintf(stderr, "Cannot disable piglow: %s\n", strerror(err));
        goto fail;
    }
    if ((err = piglow_close(&glow))) {
        fprintf(stderr, "Cannot close piglow: %s\n", strerror(err));
        goto fail;
    }
    return 0;
fail:
    piglow_close(&glow); // ignore errors
    return 1;
}
